[DOC-977][2025.1.1] Table-level locks support (#28601)
yugabyte/yugabyte-db: docs/content/stable/architecture/transactions/concurrency-control.md

PostgreSQL table locks can be broadly categorized into two types:

**Shared locks:**

- `ACCESS SHARE`
- `ROW SHARE`
- `ROW EXCLUSIVE`

**Global locks:**

- `SHARE UPDATE EXCLUSIVE`
- `SHARE`
- `SHARE ROW EXCLUSIVE`
- `EXCLUSIVE`
- `ACCESS EXCLUSIVE`

DMLs acquire shared locks alone, while DDLs acquire a combination of shared and global locks.

To reduce the overhead of DMLs, YugabyteDB takes shared locks on the PostgreSQL backend's host TServer alone. Global locks are propagated to the Master leader.

Each TServer maintains an in-memory `TSLocalLockManager` that serves object lock acquire or release calls. The Master also maintains an in-memory lock manager primarily for serializing conflicting global object locks. When the Master leader receives a global lock request, it first acquires the lock locally, then fans out the lock request to all TServers with valid [YSQL leases](#ysql-lease-mechanism), and executes the client callback after the lock has been successfully acquired on all TServers (with a valid YSQL lease).

> All statements inherently acquire some object locks. You can also explicitly acquire object locks using the [LOCK TABLE API](https://www.postgresql.org/docs/current/sql-lock.html):

Author: cc @basavaraj29

> All statements inherently acquire some object locks. You can also explicitly acquire object locks using the [LOCK TABLE API](https://www.postgresql.org/docs/current/sql-lock.html):

Reviewer: yeah, we fully support the explicit `lock table` api, and the implicit locks are takes similar to how postgres does (we tap into their lock functions and acquire the locks in a distributed manner).

can claim the same with table locks.

When a DDL finishes, all locks corresponding to the transaction are released and this release path enforces cache refresh on the the TServers ensuring that they have the latest catalog cache. This also ensures any new DMLs waiting on the same locks to see the latest schema after acquiring the object locks.

To reduce overhead for read-only workloads, YugabyteDB reuses DocDB transactions wherever possible.

### Failure handling

Locks are cleaned up for various failure scenarios as follows:

- PostgreSQL backend crash: The TServer-PostgreSQL session heartbeat cleans up corresponding locks.
- TServer crash: YSQL Lease mechanism cleans up DDL locks that originated from the TServer.
- Master leader crash: Newly elected Master leader replays catalog entries to recreate locks' state held (at the master) and establish the lease for the TServers.

### Important considerations

- If a TServer doesn't have a valid YSQL lease, it cannot serve any requests from PostgreSQL backends.
- Upon TServer lease changes (due to network issues), all active PostgreSQL backends are killed.

> - If TServers cannot communicate with the Master leader for shorter durations (lesser than YSQL lease), DDLs will stall because global locks cannot be served.

Reviewer 2: ```suggestion
- If TServers cannot communicate with the Master leader for shorter durations (less than the YSQL lease), DDLs will stall because global locks cannot be served.
```

> - If TServers cannot communicate with the Master longer than the YSQL lease timeout, they lose their YSQL lease and all PostgreSQL backends are killed (until the connection with the Master is reestablished).

Reviewer 2: ```suggestion
- If the length of time during which TServers cannot communicate with the Master exceeds the YSQL lease timeout, they lose their YSQL lease and all PostgreSQL backends are killed (until the connection with the Master is reestablished).
```
